File: Connect-four.py
'''
Assigning values to the grid
The grid will look like this:
  0,0 | 0,1 | 0,2 | 0,3 | 0,4 | 0,5 | 0,6
  1,0 | 1,1 | 1,2 | 1,3 | 1,4 | 1,5 | 1,6
  2,0 | 2,1 | 2,2 | 2,3 | 2,4 | 2,5 | 2,6
  3,0 | 3,1 | 3,2 | 3,3 | 3,4 | 3,5 | 3,6
  4,0 | 4,1 | 4,2 | 4,3 | 4,4 | 4,5 | 4,6
  5,0 | 5,1 | 5,2 | 5,3 | 5,4 | 5,5 | 5,6
'''
N, M = 6, 7
grid = [['-']*M for _ in range(N)]
#This function prints the grid of Connect Four Game as the game progresses
def print_grid():
    print("Player 1: X  vs  Player 2: O")
    print('--' + '---' * M + '--')
    for i in range(N):
        print(end='|  ')
        for j in range(M):
            print(grid[i][j], end='  ')
        print(end='|')
        print()
        print('--' + '---' * M + '--')
#This function checks if row or column or diagonal is full with same characters
def check_win(mark):
    for i in range(M-3):     #check_rows
      for j in range(N):
        if grid[j][i] == grid[j][i+1] == grid[j][i+2] == grid[j][i+3] == mark:
          return True
    for i in range(M):       #check_columns
      for j in range(N-3):
        if grid[j][i] == grid[j+1][i] == grid[j+2][i] == grid[j+3][i] == mark:
          return True
    for i in range(M-3):      #check_dagonals
      for j in range(N-3):
        if grid[j][i] == grid[j+1][i+1] == grid[j+2][i+2] == grid[j+3][i+3] == mark:
          return True
    for i in range (3,M):
      for j in range(N-3):
        if grid[j][i] == grid[j+1][i-1] == grid[j+2][i-2]==grid[j+3][i-3] == mark:
          return True
    return False
def check_grideEmpty():
    for i in range(N):
        for j in range(M):
            if grid[i][j] == '-':
                return False
    return True
#This function checks if row or column or diagonal is full with same characters
def check_tie(mark):
   if not check_grideEmpty() or check_win(mark):
       return False
   else:
       return True
#This function checks if given cell is empty or not
def check_empty(i):
    for j in range(N):
      if grid[j][i] == '-':
        return True
    return False
#This function checks if given position is valid or not
def check_valid_column(i):
    if i >= 7 or i < 0:
      return False
    else:
      return True
#This function sets a value to a cell
def set_cell(i, mark):
    j = N - 1
    while(j >= 0):
      if grid[j][i] == '-':
        grid[j][i] = mark
        break
      j -= 1
#This function clears the grid
def grid_clear():
    for i in range(M):
      for j in range(N):
        grid [j][i] = '-'
#MAIN FUNCTION
def play_game():
    print("Connect Four Game!")
    print("Welcome...")
    print("============================")
    player = 0
    while True:
        #Prints the grid
        print_grid()
        #Set mark value based on the player
        mark = 'X' if player == 0 else 'O'
        #Takes input from the user to fill in the grid
        print('Player %s' % mark)
        i = int(input('Enter the column index: '))
        while not check_valid_column(i) or not check_empty(i):
            i = int(input('Enter a valid column index: '))
        #Set the input position with the mark
        set_cell(i, mark)
        #Check if the state of the grid has a win state
        if check_win(mark):
            #Prints the grid
            print_grid()
            print('Congrats, Player %s is won!' % mark)
            break
        op_mark = 'O' if player == 0 else 'X'
        #Check if the state of the grid has a tie state
        if check_tie(op_mark):
            #Prints the grid
            print_grid()
            print("Woah! That's a tie!")
            break
        #Player number changes after each turn
        player = 1 - player
while True:
  grid_clear()
  play_game()
  c = input('Play Again [Y/N] ')
  if c not in 'yY':
      break

File: connect-four.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 6;
const int M = 7;
const int n_players = 2;
char marks[n_players] = {'X', 'O'};
char grid[N][M];
//This function prints the grid of Connect Four Game as the game progresses
void print_grid() {
	for (int i = 0; i< n_players; i++) {
        cout << "Player " << i+1 << ": " << marks[i] << "  ";
        if (i < n_players-1)
            cout << "vs  ";
	}
	cout << "\n";
	cout << "--";
    for (int i = 0; i < M; cout << "---", i++);
    cout << "--\n";
    for (int i = 0; i < N; i++) {
        cout << "|  ";
        for (int j = 0; j < M; j++)
            cout << grid[i][j] << "  ";
        cout << "|\n";
        cout << "--";
        for (int i = 0; i < M; cout << "---", i++);
        cout << "--\n";
    }
}
//This function checks if the game has a win state or not
bool check_win(char mark)
{
    for(int i = 0; i < M - 3; i++)
    {
        for(int j = 0; j < N; j++)
        {
            if (grid[j][i] == mark && grid[j][i+1] == mark && grid[j][i+2] == mark && grid[j][i+3] == mark)
                return true;
        }
    }
    for(int i = 0 ; i < M; i++)
    {
        for(int j = 0; j < N - 3; j++)
        {
            if(grid[j][i] == mark && grid[j+1][i] == mark && grid[j+2][i] == mark && grid[j+3][i] == mark)
                return true;
        }
    }
    for (int i = 0; i < M - 3; i++)
    {
        for(int j = 0; j < N - 3; j++)
        {
            if (grid[j][i] == mark && grid[j+1][i+1] == mark && grid[j+2][i+2] == mark && grid[j+3][i+3] == mark)
                return true;
        }
    }
    for (int i = 3; i < M; i++)
    {
        for(int j = 0; j < N - 3; j++)
        {
            if (grid[j][i] == mark && grid[j+1][i-1] == mark && grid[j+2][i-2] == mark && grid[j+3][i-3] == mark)
                return true;
        }
        return false;
    }
}
bool check_gridEmpty(){
  for(int i = 0; i < N; i++)
  {
    for(int j = 0; j < M; j++)
    {
      if (grid[i][j] == '-')
      return false;
    }
  }
  return true;
}
//This function checks if the game has a tie state or not
bool check_tie(char mark) {
  if(!check_gridEmpty() || check_win(mark))
  return false;
  else
  return true;
}
//This function checks if given cell is empty or not
bool check_empty(int i) {
    if(grid[0][i] == '-')
        return true;
  return false;
}
//This function checks if given position is valid or not
bool check_valid_column(int i) {
  if (i >= 0 && i <= 7)
    return true;
  else
    return false;
}
//This function sets the given mark to the given column
void set_cell(int i, char mark)
{
    int j = N - 1;
    while(j >= 0)
    {
        if (grid[j][i] == '-')
        {
            grid[j][i] = mark;
            break;
        }
        j -= 1;
    }
}
//This function clears the game structures
void grid_clear()
 {
    for(int i = 0; i < M; i++)
    {
        for(int j = 0; j < N; j++)
        {
            grid[j][i] = '-';
        }
    }
}
//This function reads a valid position input
void read_input(int &i) {
    cout << "Enter the column index: ";
    cin >> i;
    while (!check_valid_column(i) || !check_empty(i)) {
        cout << "Enter a valid column index: ";
        cin >> i;
    }
}
//MAIN FUNCTION
void play_game() {
    grid_clear();
    cout << "Connect Four Game!\n";
    cout << "Welcome...\n";
    cout << "============================\n";
    int player = 0;
    while (true) {
        //Prints the grid
        print_grid();
        //Read an input from the player
        cout << "Player " << marks[player] << " is playing now\n";
        int i;
		read_input(i);
        //Set the input position with the mark
        set_cell(i, marks[player]);
        //Check if the state of the grid has a win state
        if (check_win(marks[player])) {
            //Prints the grid
            print_grid();
            //Announcement of the final statement
            cout << "Congrats, Player " << marks[player] << " is won!\n";
            break;
        }
        //Check if the state of the grid has a tie state
        if (check_tie(marks[player])) {
            //Prints the grid
            print_grid();
            //Announcement of the final statement
            cout << "Woah! That's a tie!\n";
            break;
        }
        //Player number changes after each turn
        player = (player + 1) % n_players;
    }
}
int main() {
    while (true) {
    	grid_clear();
    	play_game();
    	char c;
    	cout << "Play Again [Y/N] ";
    	cin >> c;
    	if (c != 'y' && c != 'Y')
    		break;
    }
}
